src/Interpreters/InterpreterOptimizeQuery.cpp

```diff
@@ -27,7 +32,38 @@ BlockIO InterpreterOptimizeQuery::execute()
     auto table_id = context.resolveStorageID(ast, Context::ResolveOrdinary);
     StoragePtr table = DatabaseCatalog::instance().getTable(table_id, context);
     auto metadata_snapshot = table->getInMemoryMetadataPtr();
-    table->optimize(query_ptr, metadata_snapshot, ast.partition, ast.final, ast.deduplicate, context);
+
+    // Empty list of names means we deduplicate by all columns, but user can explicitly state which columns to use.
+    Names column_names;
+    if (ast.deduplicate_by_columns)
+    {
+        // User requested custom set of columns for deduplication, possibly with Column Transformer expression.
+        {
+            // Expand asterisk, column transformers, etc into list of column names.
+            const auto cols = processColumnTransformers(context.getCurrentDatabase(), table, metadata_snapshot, ast.deduplicate_by_columns);
+            for (const auto & col : cols->children)
+                column_names.emplace_back(col->getColumnName());
+        }
+
+        metadata_snapshot->check(column_names, NamesAndTypesList{}, table_id);
+        const auto & sorting_keys = metadata_snapshot->getColumnsRequiredForSortingKey();
+        for (const auto & sorting_key : sorting_keys)
+        {
+            // Deduplication is performed only for adjacent rows in a block,
+            // and all rows in block are in the sorting key order,
+            // hence deduplication always implicitly takes sorting keys in account.
+            // So we just explicitly state that limitation in order to avoid confusion.
+            if (std::find(column_names.begin(), column_names.end(), sorting_key) == column_names.end())
+                throw Exception(ErrorCodes::THERE_IS_NO_COLUMN,
+                        "DEDUPLICATE BY expression must include all columns used in table's ORDER BY or PRIMARY KEY,"
+                        " but '{}' is missing."
+                        " Expanded deduplicate columns expression: ['{}']",
+                        sorting_key, fmt::join(column_names, "', '"));
+        }
+    }
```

Comment:
And what's with partition key columns?

Reply:
Partition key is part of sorting key, there is also a case for the in the tests:

https://github.com/ClickHouse/ClickHouse/pull/17846/files#diff-c91002f2000111398950b55920142b1aebc6711be2f3fc195b84f79e57568f80R30

Reply:
Good catch, fixed